controller/tests/controller_server/csi_controller_server_test.py

```diff
@@ -1262,3 +1263,104 @@ def test_identity_probe(self):
+
+        self.assertEqual(self.context.code, grpc.StatusCode.OK)
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_validate_volume_cap_with_empty_id(self, storage_agent):
+        storage_agent.return_value = self.storage_agent
+        self.request.volume_id = ""
+        res = self.get_create_object_method()(self.request, self.context)
+        self.assertEqual(self.context.code, grpc.StatusCode.INVALID_ARGUMENT)
+        self.assertTrue("volume id" in self.context.details)
+        self.assertEqual(res, self.get_create_object_response_method()(message=self.context.details))
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_validate_volume_cap_with_wrong_secrets(self, storage_agent):
+        self._test_create_object_with_wrong_secrets(storage_agent)
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_validate_volume_cap_with_unsupported_capabilities(self, storage_agent):
+        storage_agent.return_value = self.storage_agent
+        self.request.volume_capabilities[0].access_mode.mode = 999
+        res = self.get_create_object_method()(self.request, self.context)
+        self.assertEqual(self.context.code, grpc.StatusCode.INVALID_ARGUMENT)
+        self.assertTrue("unsupported access mode" in self.context.details)
+        self.assertEqual(res, self.get_create_object_response_method()(message=self.context.details))
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_validate_volume_cap_with_no_capabilities(self, storage_agent):
+        storage_agent.return_value = self.storage_agent
+        self.request.volume_capabilities = {}
+        res = self.get_create_object_method()(self.request, self.context)
+        self.assertEqual(self.context.code, grpc.StatusCode.INVALID_ARGUMENT)
+        self.assertTrue("not set" in self.context.details)
+        self.assertEqual(res, self.get_create_object_response_method()(message=self.context.details))
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_validate_volume_cap_with_bad_id(self, storage_agent):
+        storage_agent.return_value = self.storage_agent
+        self.request.volume_id = "wwn1"
+        res = self.get_create_object_method()(self.request, self.context)
+        self.assertEqual(self.context.code, grpc.StatusCode.NOT_FOUND)
+        self.assertTrue("id format" in self.context.details)
+        self.assertEqual(res, self.get_create_object_response_method()(message=self.context.details))
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_validate_volume_cap_with_vol_not_found(self, storage_agent):
+        storage_agent.return_value = self.storage_agent
+        self.mediator.get_volume_by_id.side_effect = [array_errors.VolumeNotFoundError("vol")]
+        res = self.get_create_object_method()(self.request, self.context)

```

Comment:
IMO it is better to leave 1 blank line between each part of the unit test (BUILD-OPERATE-CHECK/ARRANGE-ACT-ASSERT/etc...)

Reply:
done

Reply:
nice!